[AMDGPU] Insert waitcnt for non-global fence release in GFX12

ritter-x2a opened the pull request:
A fence release could be followed by a barrier, so it should wait for
the relevant memory accesses to complete, even if it is mmra-limited to
LDS. So far, that would be skipped for non-global fence releases.

Fixes SWDEV-554932.

ritter-x2a committed:
[AMDGPU] Insert waitcnt for non-global fence release in GFX12

A fence release could be followed by a barrier, so it should wait for
the relevant memory accesses to complete, even if it is mmra-limited to
LDS. So far, that would be skipped for non-global fence releases.

Fixes SWDEV-554932.

ritter-x2a commented:
* **#159282** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/159282?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/159282?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2522,8 +2522,7 @@ bool SIGfx12CacheControl::insertRelease(MachineBasicBlock::iterator &MI,
-  if ((AddrSpace & SIAtomicAddrSpace::GLOBAL) == SIAtomicAddrSpace::NONE)
+  if (AddrSpace == SIAtomicAddrSpace::SCRATCH)
     return false;
 
```
I think this can be fixed more easily by moving all of the code surrounding the switch below into something like `if((AddrSpace & SIAtomicAddrSpace::GLOBAL) != SIAtomicAddrSpace::NONE || ((AddrSpace & SIAtomicAddrSpace::SCRATCH)  != SIAtomicAddrSpace::NONE) && ST.hasGloballyAddressableScratch())`

Basically always go to the InsertWait section, just skip the inv insertion if we're not touching the global AS.
That way we don't get those global_wb inserted for the local fences in the test below

ssahasra replied:
The original code had an early return on a negative condition which matched the one in `insertAcquire()`. Why change it and move all the code inside a positive condition? That seems to have introduced a lot of whitespace noise due to reindentation.

ssahasra replied:
I see now ... I had not noticed the `insertWait()` at the end. `canAffectGlobalAS()` seems like a good idea. And maybe we can also have `insertWB()`, so that it's clear what is happening at a release? In either case, I think it's safer to fix that early return as a separate commit so that any breakages can be investigated separately.

ritter-x2a replied:
> And maybe we can also have insertWB(), so that it's clear what is happening at a release?

Feel free to overrule me here, but I don't think we gain much from introducing `insertWB`. I find that logic only used in three `insertRelease` implementations (gfx90a, gfx940, and gfx12), where it does something slightly different in each case, so there isn't much duplication to get rid off.
We could turn the `insertRelease` implementations for gfx90a, gfx940, and gfx12 into `insertWB(...); insertWait(...);` (for the others it seems to be just `insertWait(...);`), but that would only move the WB code from `SIGfx*CacheControl::insertRelease` to the new `SIGfx*CacheControl::insertWB`.

ritter-x2a replied:
I've separated (and adjusted, to reflect my above comment about the GloballyAddressableScratch part being unnecessary) the part that concerned GloballyAddressableScratch into #160129.

Pierre-vh reviewed (commented):
Can you also check if other CacheControl classes have a similar issue, and if the acquire equivalent has a similar issue too (first check if you can reproduce a similar issue in a test case) ?

MMRAs touch the `OrderingAddrSpace`, so anything that checks for it too eagerly could be fooled unfortunately

krzysz00 commented on llvm/test/CodeGen/AMDGPU/memory-legalizer-fence-mmra-local.ll:
```diff
@@ -810,14 +827,18 @@ define amdgpu_kernel void @agent_seq_cst_fence() {
 ;
 ; GFX1250-LABEL: agent_seq_cst_fence:
 ; GFX1250:       ; %bb.0: ; %entry
+; GFX1250-NEXT:    global_wb scope:SCOPE_DEV
```
Re review comments: ... Wait, isn't a device-scope cache invalidation part of a device-scope release fence though?

kuhar reviewed (commented):
Avoid internal ticket URLs in PR descriptions: https://github.com/llvm/llvm-project/commit/677326999f270395ecaec026178fe7ed148a0068

arsenm commented:
> [6773269](https://github.com/llvm/llvm-project/commit/677326999f270395ecaec026178fe7ed148a0068)

The policy was designed to specifically allow this

ritter-x2a committed:
Don't emit WB for local-AS fences...

...and treat release/acquire on GloballyAddressableScratch as global.

This also fixes an issue where an early return between an iterator
increment/decrement pair in SIGfx12CacheControl::insertRelease could have lead
to unexpected insertion locations.

ritter-x2a committed:
Remove changes related to GloballyAddressableScratch from this PR

ritter-x2a committed:
Properly report changes in SIGfx12CacheControl::insertRelease
